File: src/harness/opencode.rs
// Copyright 2026 Joe Drago. All rights reserved.
// SPDX-License-Identifier: BSD-2-Clause

//! opencode harness: writes `~/.config/opencode/opencode.{jsonc,json}` and
//! manages the `permission.bash` curl allowlist for lui's web-search port.
//! Respects `harness_opencode_disable_prune` when set.

use jsonc_parser::cst::{CstInputValue, CstObject, CstObjectProp, CstRootNode};
use jsonc_parser::ParseOptions;
use regex::Regex;

use super::{arr, b, n, obj, s, ConfigFile, Harness, HarnessInputs};
use crate::settings::store::Effective;
use crate::ssh_tunnel::{ssh_run, SshTarget};

pub const HARNESS: Harness = Harness {
    name: "opencode",
    setting_name: "harness_opencode",
    flag_long: "harness-opencode",
    default_on: true,
    help: &[
        "Manage opencode.{jsonc,json} and the lui-web-search skill",
        "Leave opencode's config alone (for external harnesses)",
    ],
    config: ConfigFile {
        dir: ".config/opencode",
        candidates: &["opencode.jsonc", "opencode.json"],
    },
    apply,
    needs_backup,
    preflight_ssh: Some(preflight_ssh),
};

fn apply(root: &CstObject, eff: &Effective, inputs: &HarnessInputs) {
    set_provider_lui(root, &inputs.model_name, &inputs.base_url, inputs.ctx_size);
    set_permission_bash(root, inputs.web_port, inputs.websearch);
    if eff
        .get_bool("harness_opencode_disable_prune")
        .unwrap_or(false)
    {
        set_compaction_prune_false_if_absent(root);
    }
}

/// True iff the text has content but no `provider.lui` entry. Decides
/// whether to drop a `.luibackup` before writing. Empty/absent → false
/// (nothing to back up). Unparseable → true (preserve it).
fn needs_backup(existing: &str) -> bool {
    if existing.trim().is_empty() {
        return false;
    }
    let Ok(root) = CstRootNode::parse(existing, &ParseOptions::default()) else {
        return true;
    };
    let Some(root_obj) = root.object_value() else {
        return true;
    };
    let Some(provider) = root_obj.object_value("provider") else {
        return true;
    };
    provider.get("lui").is_none()
}

/// Verify `opencode` is installed on the remote before we start editing
/// its config. Non-interactive SSH runs a non-login shell, so PATH edits
/// the opencode installer makes in `~/.bashrc` / `~/.zshrc` aren't
/// sourced. We OR three checks together so any single success suffices:
/// `command -v opencode` (default PATH), `bash -lc 'command -v opencode'`
/// (login-shell PATH — picks up `~/.bash_profile` / `~/.profile` edits),
/// and the installer's canonical location `~/.opencode/bin/opencode`.
/// Any one succeeding means opencode is usable once the user SSHes in.
fn preflight_ssh(target: &SshTarget) -> Result<(), String> {
    let probe = "command -v opencode \
        || bash -lc 'command -v opencode' \
        || { [ -x \"$HOME/.opencode/bin/opencode\" ] && echo \"$HOME/.opencode/bin/opencode\"; }";
    match ssh_run(target, &[probe], None) {
        Ok(out) if !out.trim().is_empty() => Ok(()),
        Ok(_) | Err(_) => Err(format!(
            "opencode not found on {}. Install it there first.",
            target.spec()
        )),
    }
}

fn set_provider_lui(root_obj: &CstObject, model_name: &str, llama_base_url: &str, ctx_size: u32) {
    let provider = root_obj.object_value_or_set("provider");

    let model_entry = CstInputValue::Object(vec![(
        model_name.to_string(),
        obj([
            ("name", s(model_name)),
            ("supportsToolCalls", b(true)),
            (
                "limit",
                obj([
                    ("context", n(ctx_size)),
                    ("input", n(ctx_size)),
                    ("output", n(8192)),
                ]),
            ),
        ]),
    )]);

    let lui_value = obj([
        ("name", s("lui")),
        ("npm", s("@ai-sdk/openai-compatible")),
        (
            "options",
            obj([
                ("baseURL", s(llama_base_url)),
                (
                    "toolParser",
                    arr(vec![
                        obj([("type", s("raw-function-call"))]),
                        obj([("type", s("json"))]),
                    ]),
                ),
            ]),
        ),
        ("models", model_entry),
    ]);

    match provider.get("lui") {
        Some(prop) => prop.set_value(lui_value),
        None => {
            provider.append("lui", lui_value);
        }
    }
}

/// Manage `permission.bash["curl*http://127.0.0.1:<port>/*"]` so opencode
/// doesn't prompt for every search curl. Cleans stale port wildcards so a
/// `--web-port` change doesn't leave a dead allowlist entry behind. When
/// websearch is disabled we refuse to create parent objects (`permission`
/// / `bash`) just to insert nothing.
fn set_permission_bash(root_obj: &CstObject, web_port: u16, websearch: bool) {
    let current_pattern = format!("curl*http://127.0.0.1:{}/*", web_port);

    let has_permission = root_obj.get("permission").is_some();
    if !websearch && !has_permission {
        return;
    }
    let permission = root_obj.object_value_or_set("permission");

    let has_bash = permission.get("bash").is_some();
    if !websearch && !has_bash {
        return;
    }
    let bash = permission.object_value_or_set("bash");

    let stale_re = Regex::new(r"^curl\*http://127\.0\.0\.1:\d+/\*$").unwrap();
    let to_remove: Vec<CstObjectProp> = bash
        .properties()
        .into_iter()
        .filter(|p| {
            let Some(name) = p.name() else { return false };
            let Ok(key) = name.decoded_value() else {
                return false;
            };
            stale_re.is_match(&key) && key != current_pattern
        })
        .collect();
    for prop in to_remove {
        prop.remove();
    }

    if !websearch {
        if let Some(prop) = bash.get(&current_pattern) {
            prop.remove();
        }
    } else {
        match bash.get(&current_pattern) {
            Some(prop) => prop.set_value(s("allow")),
            None => {
                bash.append(&current_pattern, s("allow"));
            }
        }
    }
}

/// Set `compaction.prune = false` only if the key is absent — never
/// clobbers a user-set value. Creates the parent `compaction` object
/// only if missing.
fn set_compaction_prune_false_if_absent(root_obj: &CstObject) {
    let compaction = root_obj.object_value_or_set("compaction");
    if compaction.get("prune").is_none() {
        compaction.append("prune", b(false));
    }
}

File: src/harness/opencode.js
// opencode harness: writes ~/.config/opencode/opencode.{jsonc,json}.
// Inserts a `provider.lui` block pointing opencode at llama-server,
// and a `permission.bash` curl allowlist for lui's web-search port.
//
// `reasoning_effort` (unset by default) writes
// provider.lui.models.<model>.options.reasoningEffort. opencode feeds
// model options to the AI SDK, whose openai-compatible provider maps
// reasoningEffort to the request body's `reasoning_effort`, which
// llama-server hands to the chat template. Qwen3.8 accepts low,
// medium, xhigh; `none` disables thinking entirely.

/** @import { Harness } from "../types.js" */

import { modify, applyEdits, parseTree, findNodeAtLocation } from "jsonc-parser"

const FORMAT = { tabSize: 2, insertSpaces: true, eol: "\n" }

/** @type {Harness} */
export const harness = {
    name: "opencode",
    configDir: "~/.config/opencode",
    configCandidates: ["opencode.jsonc", "opencode.json"],
    skillsDir: "skills",
    schema: [
        { path: "enabled", default: false },
        { path: "reasoning_effort", default: null }
    ],

    apply(existing, ctx, config) {
        let text = existing.trim() ? existing : "{}\n"
        const effort = typeof config.reasoning_effort === "string" && config.reasoning_effort ? config.reasoning_effort : null
        text = setProviderLui({
            text,
            modelName: ctx.modelName,
            baseURL: ctx.baseURL,
            ctxSize: ctx.ctxSize,
            maxOutputTokens: ctx.maxOutputTokens,
            reasoningEffort: effort
        })
        text = setPermissionBash(text, ctx.webPort, ctx.websearch)
        return text
    },

    needsBackup(existing) {
        if (!existing.trim()) return false
        const tree = parseTree(existing, [], { allowTrailingComma: true })
        if (!tree) return true
        const lui = findNodeAtLocation(tree, ["provider", "lui"])
        return !lui
    },

    async sshPreflight(remote) {
        try {
            if (await remote.which("opencode")) return { ok: true }
            // opencode's installer drops the binary here and edits the
            // shell profile to add it, which a non-interactive session
            // never reads — so check the install path directly.
            const installed = remote.platform === "win32" ? "~/.opencode/bin/opencode.exe" : "~/.opencode/bin/opencode"
            if (await remote.exists(installed)) return { ok: true }
            return { ok: false, error: `opencode not found on ${remote.spec}. Install it there first.` }
        } catch (e) {
            return {
                ok: false,
                error: `opencode preflight on ${remote.spec} failed: ${/** @type {Error} */ (e).message}`
            }
        }
    }
}

/** @param {{ text: string, modelName: string, baseURL: string, ctxSize: number, maxOutputTokens: number, reasoningEffort: string | null }} args @returns {string} */
function setProviderLui({ text, modelName, baseURL, ctxSize, maxOutputTokens, reasoningEffort }) {
    /** @type {{ name: string, supportsToolCalls: boolean, limit: { context: number, input: number, output: number }, options?: { reasoningEffort: string } }} */
    const modelValue = {
        name: modelName,
        supportsToolCalls: true,
        limit: { context: ctxSize, input: ctxSize, output: maxOutputTokens }
    }
    if (reasoningEffort) modelValue.options = { reasoningEffort }
    const luiValue = {
        name: "lui",
        npm: "@ai-sdk/openai-compatible",
        options: {
            baseURL,
            toolParser: [{ type: "raw-function-call" }, { type: "json" }]
        },
        models: {
            [modelName]: modelValue
        }
    }
    const edits = modify(text, ["provider", "lui"], luiValue, { formattingOptions: FORMAT })
    return applyEdits(text, edits)
}

/** @param {string} text @param {number} webPort @param {boolean} websearch @returns {string} */
function setPermissionBash(text, webPort, websearch) {
    const currentPattern = `curl*http://127.0.0.1:${webPort}/*`
    let cur = text

    // Sweep stale loopback patterns at any port, keeping the current one.
    const tree = parseTree(cur, [], { allowTrailingComma: true })
    const stale = /^curl\*http:\/\/127\.0\.0\.1:\d+\/\*$/
    const bashNode = tree && findNodeAtLocation(tree, ["permission", "bash"])
    if (bashNode?.type === "object") {
        for (const prop of bashNode.children ?? []) {
            const key = prop.children?.[0]
            if (!key || typeof key.value !== "string") continue
            if (key.value === currentPattern) continue
            if (!stale.test(key.value)) continue
            const edits = modify(cur, ["permission", "bash", key.value], undefined, { formattingOptions: FORMAT })
            cur = applyEdits(cur, edits)
        }
    }

    if (websearch) {
        const edits = modify(cur, ["permission", "bash", currentPattern], "allow", { formattingOptions: FORMAT })
        cur = applyEdits(cur, edits)
    } else {
        // jsonc-parser's modify(..., undefined) throws on a path that
        // doesn't already exist (e.g. fresh install with no permission
        // block). Only attempt the delete when the entry is actually
        // there.
        const post = parseTree(cur, [], { allowTrailingComma: true })
        if (post && findNodeAtLocation(post, ["permission", "bash", currentPattern])) {
            const edits = modify(cur, ["permission", "bash", currentPattern], undefined, { formattingOptions: FORMAT })
            cur = applyEdits(cur, edits)
        }
    }
    return cur
}
